### PR TITLE
std: Add `arch` and `simd` modules

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit imports the `stdsimd` crate into the standard library,
creating an `arch` and `simd` module inside of both libcore and libstd.
Both of these modules are **unstable** and will continue to be so until
RFC 2335 is stabilized.

As a brief recap, the modules are organized as so:

* `arch` contains all current architectures with intrinsics, for example
  `std::arch::x86`, `std::arch::x86_64`, `std::arch::arm`, etc. These
  modules contain all of the intrinsics defined for the platform, like
  `_mm_set1_epi8`.
* In the standard library, the `arch` module also exports a
  `is_target_feature_detected` macro which performs runtime detection to
  determine whether a target feature is available at runtime.
* The `simd` module contains experimental versions of strongly-typed
  lane-aware SIMD primitives, to be fully fleshed out in a future RFC.

The main purpose of this commit is to start pulling in all these
intrinsics and such into the standard library on nightly and allow
testing and such. This'll help allow users to easily kick the tires and
see if intrinsics work as well as allow us to test out all the
infrastructure for moving the intrinsics into the standard library.

### CODE REVIEW

File: src/libcore/lib.rs [hunk 1]
```diff
@@ -68,31 +68,39 @@
...
 #![feature(specialization)]
 #![feature(staged_api)]
+#![feature(stmt_expr_attributes)]
+#![feature(target_feature)]
 #![feature(unboxed_closures)]
 #![feature(untagged_unions)]
 #![feature(unwind_attributes)]
-#![feature(doc_spotlight)]
-#![feature(rustc_const_unstable)]
-#![feature(iterator_repeat_with)]
```
Comment (joshtriplett) [inline src/libcore/lib.rs]: Minor nit, but shouldn't moving these three around into a different spot be in a separate commit?

**Comment by joshtriplett** [comment]
One minor nit, which I don't feel particularly strongly about, and otherwise this LGTM.

**Comment by alexcrichton** [comment]
cc @rust-lang/libs, @BurntSushi, @gnzlbg

**Comment by sfackler** [comment]
Is the plan to always be importing stdsimd, or is this just temporary and it'll get inlined into this repo at some point?

**Comment by alexcrichton** [comment]
My thinking at least is that we'll forever be importing stdsimd as a submodule, that allows us to be way more creative with stdsimd's own CI and doesn't require us to hook it all up here as well

File: src/libcore/lib.rs [hunk 2]
```diff
@@ -177,3 +185,21 @@ mod char_private;
...
+// Pull in the the `coresimd` crate directly into libcore. This is where all the
+// architecture-specific (and vendor-specific) intrinsics are defined. AKA
+// things like SIMD and such. Note that the actual source for all this lies in a
+// different repository, rust-lang-nursery/stdsimd. That's why the setup here is
+// a bit wonky.
+#[path = "../stdsimd/coresimd/mod.rs"]
+#[allow(missing_docs, missing_debug_implementations, dead_code)]
+#[unstable(feature = "stdsimd", issue = "0")]
+#[cfg(not(stage0))] // allow changes to how stdsimd works in stage0
+mod coresimd;
```
Comment (scottmcm) [inline src/libcore/lib.rs]: Silly question: does this stage0 check preclude core/std from using anything from stdsimd?

Comment (alexcrichton) [reply src/libcore/lib.rs]: It would during stage0, yeah, but I'd imagine that most usage would largely be an optimization in which case tagging it as `#[cfg(not(stage0))]` probably wouldn't be too bad.

Comment (alexcrichton) [reply src/libcore/lib.rs]: Note that the alternative here is to manually add `#[cfg(stage0)]` tags and such to stdsimd itself rather than blanket defining it for this module, and I figured that for the time being it's easiest to take this route which is only defining it in stage1+

Comment (scottmcm) [reply src/libcore/lib.rs]: Oh, good point.  Yeah, the stage1 compiler just being a bit slower in some corner cases than it could be probably isn't a big deal.

**Comment by gnzlbg** [comment]
LGTM. 

Pulling in the `simd` module was a bit unexpected, but at least this will tell us if this really does work on all targets supported by rustc.

**Comment by alexcrichton** [comment]
@gnzlbg do you think the `simd` module should be left out for now?

**Comment by gnzlbg** [comment]
Nah, it's unstable.

**Comment by scottmcm** [comment]
FWIW I'd like to use `u8x{8, 16, 32}` in core, so it'd be great to have `simd`.

**Comment by gnzlbg** [comment]
@scottmcm just be aware that the API will change a bit shortly and you'll be ok.

**Comment by kennytm** [comment]
@bors r-

The `bmi` target feature has been renamed to `bmi1` since #48565. Please rebase and update the `stdsimd` submodule.

<details>

```
[00:26:13] error: the feature named `bmi` is not valid for this target
[00:26:13]   --> libcore/../stdsimd/coresimd/x86/bmi.rs:18:18
[00:26:13]    |
[00:26:13] 18 | #[target_feature(enable = "bmi")]
[00:26:13]    |                  ^^^^^^^^^^^^^^
[00:26:13] 
[00:26:13] error: the feature named `bmi` is not valid for this target
[00:26:13]   --> libcore/../stdsimd/coresimd/x86/bmi.rs:30:18
[00:26:13]    |
[00:26:13] 30 | #[target_feature(enable = "bmi")]
[00:26:13]    |                  ^^^^^^^^^^^^^^
[00:26:13] 
[00:26:13] error: the feature named `bmi` is not valid for this target
[00:26:13]   --> libcore/../stdsimd/coresimd/x86/bmi.rs:38:18
[00:26:13]    |
[00:26:13] 38 | #[target_feature(enable = "bmi")]
[00:26:13]    |                  ^^^^^^^^^^^^^^
[00:26:13] 
[00:26:13] error: the feature named `bmi` is not valid for this target
[00:26:13]   --> libcore/../stdsimd/coresimd/x86/bmi.rs:46:18
[00:26:13]    |
[00:26:13] 46 | #[target_feature(enable = "bmi")]
[00:26:13]    |                  ^^^^^^^^^^^^^^
[00:26:13] 
[00:26:13] error: the feature named `bmi` is not valid for this target
[00:26:13]   --> libcore/../stdsimd/coresimd/x86/bmi.rs:54:18
[00:26:13]    |
[00:26:13] 54 | #[target_feature(enable = "bmi")]
[00:26:13]    |                  ^^^^^^^^^^^^^^
[00:26:13] 
[00:26:13] error: the feature named `bmi` is not valid for this target
[00:26:13]   --> libcore/../stdsimd/coresimd/x86/bmi.rs:64:18
[00:26:13]    |
[00:26:13] 64 | #[target_feature(enable = "bmi")]
[00:26:13]    |                  ^^^^^^^^^^^^^^
[00:26:13] 
[00:26:13] error: the feature named `bmi` is not valid for this target
[00:26:13]   --> libcore/../stdsimd/coresimd/x86/bmi.rs:74:18
[00:26:13]    |
[00:26:13] 74 | #[target_feature(enable = "bmi")]
[00:26:13]    |                  ^^^^^^^^^^^^^^
[00:26:13] 
[00:26:13] error: the feature named `bmi` is not valid for this target
[00:26:13]   --> libcore/../stdsimd/coresimd/x86/bmi.rs:84:18
[00:26:13]    |
[00:26:13] 84 | #[target_feature(enable = "bmi")]
[00:26:13]    |                  ^^^^^^^^^^^^^^
[00:26:13] 
[00:26:13] error: the feature named `bmi` is not valid for this target
[00:26:13]   --> libcore/../stdsimd/coresimd/x86_64/bmi.rs:18:18
[00:26:13]    |
[00:26:13] 18 | #[target_feature(enable = "bmi")]
[00:26:13]    |                  ^^^^^^^^^^^^^^
[00:26:13] 
[00:26:13] error: the feature named `bmi` is not valid for this target
[00:26:13]   --> libcore/../stdsimd/coresimd/x86_64/bmi.rs:31:18
[00:26:13]    |
[00:26:13] 31 | #[target_feature(enable = "bmi")]
[00:26:13]    |                  ^^^^^^^^^^^^^^
[00:26:13] 
[00:26:13] error: the feature named `bmi` is not valid for this target
[00:26:13]   --> libcore/../stdsimd/coresimd/x86_64/bmi.rs:40:18
[00:26:13]    |
[00:26:13] 40 | #[target_feature(enable = "bmi")]
[00:26:13]    |                  ^^^^^^^^^^^^^^
[00:26:13] 
[00:26:13] error: the feature named `bmi` is not valid for this target
[00:26:13]   --> libcore/../stdsimd/coresimd/x86_64/bmi.rs:48:18
[00:26:13]    |
[00:26:13] 48 | #[target_feature(enable = "bmi")]
[00:26:13]    |                  ^^^^^^^^^^^^^^
[00:26:13] 
[00:26:13] error: the feature named `bmi` is not valid for this target
[00:26:13]   --> libcore/../stdsimd/coresimd/x86_64/bmi.rs:57:18
[00:26:13]    |
[00:26:13] 57 | #[target_feature(enable = "bmi")]
[00:26:13]    |                  ^^^^^^^^^^^^^^
[00:26:13] 
[00:26:13] error: the feature named `bmi` is not valid for this target
[00:26:13]   --> libcore/../stdsimd/coresimd/x86_64/bmi.rs:68:18
[00:26:13]    |
[00:26:13] 68 | #[target_feature(enable = "bmi")]
[00:26:13]    |                  ^^^^^^^^^^^^^^
[00:26:13] 
[00:26:13] error: the feature named `bmi` is not valid for this target
[00:26:13]   --> libcore/../stdsimd/coresimd/x86_64/bmi.rs:79:18
[00:26:13]    |
[00:26:13] 79 | #[target_feature(enable = "bmi")]
[00:26:13]    |                  ^^^^^^^^^^^^^^
[00:26:13] 
[00:26:13] error: the feature named `bmi` is not valid for this target
[00:26:13]   --> libcore/../stdsimd/coresimd/x86_64/bmi.rs:89:18
[00:26:13]    |
[00:26:13] 89 | #[target_feature(enable = "bmi")]
[00:26:13]    |                  ^^^^^^^^^^^^^^
[00:26:13] 
[00:26:15]    Compiling rustc_tsan v0.0.0 (file:///checkout/src/librustc_tsan)
[00:26:15]    Compiling rustc_lsan v0.0.0 (file:///checkout/src/librustc_lsan)
[00:26:17]    Compiling rustc_msan v0.0.0 (file:///checkout/src/librustc_msan)
[00:26:17]    Compiling rustc_asan v0.0.0 (file:///checkout/src/librustc_asan)
[00:26:37] error: aborting due to 16 previous errors
[00:26:37] 
[00:26:37] error: Could not compile `core`.
```

</details>

**Comment by Manishearth** [comment]
```
[00:48:53] error: function is never used: `set_bit`
[00:48:53]   --> libstd/../stdsimd/stdsimd/arch/detect/cache.rs:13:1
[00:48:53]    |
[00:48:53] 13 | pub const fn set_bit(x: u64, bit: u32) -> u64 {
[00:48:53]    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[00:48:53]    |
[00:48:53] note: lint level defined here
[00:48:53]   --> libstd/lib.rs:232:31
[00:48:53]    |
[00:48:53] 232| #![cfg_attr(not(stage0), deny(warnings))]
[00:48:53]    |                               ^^^^^^^^
[00:48:53]    = note: #[deny(dead_code)] implied by #[deny(warnings)]
[00:48:53] 
[00:48:53] error: method is never used: `set`
[00:48:53]   --> libstd/../stdsimd/stdsimd/arch/detect/cache.rs:48:5
[00:48:53]    |
[00:48:53] 48 |     pub fn set(&mut self, bit: u32) {
[00:48:53]    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[00:48:53] 
[00:48:53] error: aborting due to 2 previous errors
```
@bors r-

**Comment by alexcrichton** [comment]
@Manishearth thanks! W/ an r- mind throwing in a link to the error logs as well? In [this case](https://travis-ci.org/rust-lang/rust/jobs/348404267#L5531) it's important that it's wasm, for example, and doesn't otherwise happen by default.

**Comment by gnzlbg** [comment]
@Manishearth is there a way to view the result of all build bots?

**Comment by Manishearth** [comment]
https://travis-ci.org/rust-lang/rust/builds/348404227?utm_source=github_status&utm_medium=notification
